### ISSUE TITLE
Error message not shown on API error response for undoable actions with handled side effects.

### ISSUE THREAD

**Issue by tusmenko** [marmelab/react-admin#3244]
**What you were expecting:**
As mentioned in **Optimistic Rendering and Undo** part of documentation:
> If the fetch fails, react-admin shows an error notification, and forces a refresh, too.

Implementing Component that used basic (UPDATE in my case) actions using `withDataProvider`, and setting `onSuccess` and `onError` side effects handlers I expect to see corresponding error and success messages from API\dataProvider.

**What happened instead:**
Only message defined in `onSuccess` appears on click, "optimistically". 
But no Error message shown on API response with error, even if list view updates and optimistically rendered value falls back to initial state.

**Steps to reproduce:**
- Go to sandbox example
- On Posts tab click the custom Status switcher
-> Switch optimistically toggles it's value.
- After mocked dataProvider throws HttpError with error message, switch toggles back.
-> NO error message shown.

**Related code:**
https://codesandbox.io/s/47t8x

Consider my code modification on:
1) src/posts/StatusSwitch.js - custom Toggle switch component.
2) src/dataProvider.js - exception throwing on posts UPDATE.
3) src/posts/PostList.js - StatusSwitch added to row.

Custom switch powered withDataProvider configured with following options:

```js
dataProvider(
      UPDATE,
      resource,
      {
        id: record.id,
        data: updatedRecord
      },
      {
        undoable: true,
        onSuccess: {
          notification: { body: `${label} has changed` }
        },
        onError: {
          notification: {
            body: `Error: ${label} has not changed`,
            level: "warning"
          }
        }
      }
    );
```

**Environment**

* React-admin version: 2.8.6
* React version:16.8.6
* Browser:Chrome

**Comment by Kmaschta** [comment]
You're right, I see two problems with your CodeSandbox :

1. After optimistic update, the real request doesn't show a error notification (but reverts the change, that is ok)
2. When clicking again on status toggle, the state doesn't change. I suspect an agressive anti re-render rule on that

I'll let core maintainers check that bug.
Note : Open the CodeSandbox on full screen to see the status toggle.

**Comment by alexrosabr** [comment]
We had a similar problem here: 

1. in the List view, click Edit
2. in the Edit view, edit a field that appears in the List view and Save
3. the List view is updated with the change, 
4. after the undo delay, the API is called and returns HTTP error
5. the error message from API is correctly shown 
6. the List view is NOT updated to reverse the change

It worked on 2.9.x

**Comment by alexrosabr** [comment]
React Admin 2.9
![ezgif-ra29](https://user-images.githubusercontent.com/18428330/70828108-7d893f00-1dc9-11ea-9f37-5600f0de8e4d.gif)

React Admin 3.0
![ezgif-ra30](https://user-images.githubusercontent.com/18428330/70828128-8548e380-1dc9-11ea-9dc2-037a0026cfc6.gif)

**Comment by JulienMattiussi** [comment]
This issue has been solved by #4179
